File: dsa-tracker-backend/src/controllers/revisionController.js
const Revision = require('../models/Revision');
const UserProgress = require('../models/UserProgress');

// @desc    Get revision schedule
// @route   GET /api/revision
// @access  Private
const getRevisionSchedule = async (req, res) => {
    try {
        const { status } = req.query; // 'pending', 'completed', 'overdue'

        let query = { userId: req.user._id };

        if (status === 'completed') {
            query.completed = true;
        } else if (status === 'pending') {
            query.completed = false;
        } else if (status === 'overdue') {
            query.completed = false;
            query.scheduledDate = { $lt: new Date() };
        }

        const revisions = await Revision.find(query)
            .populate('problemId')
            .sort({ scheduledDate: 1 });

        res.json(revisions);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Get overdue revisions
// @route   GET /api/revision/overdue
// @access  Private
const getOverdueRevisions = async (req, res) => {
    try {
        const today = new Date();
        today.setHours(0, 0, 0, 0);

        const overdueRevisions = await Revision.find({
            userId: req.user._id,
            scheduledDate: { $lt: today },
            completed: false,
        })
            .populate('problemId')
            .sort({ scheduledDate: 1 });

        res.json(overdueRevisions);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Mark revision as complete
// @route   PUT /api/revision/:id/complete
// @access  Private
const markRevisionComplete = async (req, res) => {
    try {
        const revision = await Revision.findById(req.params.id);

        if (revision && revision.userId.toString() === req.user._id.toString()) {
            revision.completed = true;
            revision.completedAt = Date.now();

            await revision.save();

            // Update user progress revision count
            const progress = await UserProgress.findOne({
                userId: req.user._id,
                problemId: revision.problemId,
            });

            if (progress) {
                progress.revisionCount += 1;
                progress.revisionDates.push(Date.now());

                // Schedule next revision if applicable
                if (revision.revisionNumber < 3) {
                    const revisionDays = [1, 7, 30];
                    const nextDay = revisionDays[revision.revisionNumber];
                    const nextRevisionDate = new Date();
                    nextRevisionDate.setDate(nextRevisionDate.getDate() + nextDay);

                    progress.nextRevisionDate = nextRevisionDate;
                }

                await progress.save();
            }

            res.json(revision);
        } else {
            res.status(404).json({ message: 'Revision not found' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Get revision statistics
// @route   GET /api/revision/stats
// @access  Private
const getRevisionStats = async (req, res) => {
    try {
        const today = new Date();
        today.setHours(0, 0, 0, 0);
        const tomorrow = new Date(today);
        tomorrow.setDate(tomorrow.getDate() + 1);

        // 1. Revisions Due Today (and strictly today, not just overdue)
        // Adjusting logic: Due Today usually includes Overdue in user's mind, but let's separate them as requested.
        // Due Today: scheduledDate >= today && scheduledDate < tomorrow && !completed
        // Overdue: scheduledDate < today && !completed

        const dueTodayCount = await Revision.countDocuments({
            userId: req.user._id,
            scheduledDate: { $gte: today, $lt: tomorrow },
            completed: false
        });

        const overdueCount = await Revision.countDocuments({
            userId: req.user._id,
            scheduledDate: { $lt: today },
            completed: false
        });

        // 3. Total Problems Marked for Revision (Total active revisions pending)
        const totalPendingCount = await Revision.countDocuments({
            userId: req.user._id,
            completed: false
        });

        // 4. Weak Topics
        // We can infer weak topics from UserProgress where markedForRevision is true or revisionCount is high
        // Let's aggregate UserProgress to find topics with most active revisions
        const weakTopicsAggregation = await UserProgress.aggregate([
            {
                $match: {
                    userId: req.user._id,
                    markedForRevision: true
                }
            },
            {
                $lookup: {
                    from: 'problems',
                    localField: 'problemId',
                    foreignField: '_id',
                    as: 'problem'
                }
            },
            { $unwind: '$problem' },
            { $unwind: '$problem.topics' },
            {
                $group: {
                    _id: '$problem.topics',
                    count: { $sum: 1 }
                }
            },
            { $sort: { count: -1 } },
            { $limit: 3 }
        ]);

        const weakTopics = weakTopicsAggregation.map(t => ({ topic: t._id, count: t.count }));

        res.json({
            dueToday: dueTodayCount,
            overdue: overdueCount,
            totalPending: totalPendingCount,
            weakTopics
        });
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

// @desc    Delete a revision (Stop revising)
// @route   DELETE /api/revision/:id
// @access  Private
const deleteRevision = async (req, res) => {
    try {
        const revision = await Revision.findById(req.params.id);

        if (revision && revision.userId.toString() === req.user._id.toString()) {
            // Also update the UserProgress to stop future revisions for this problem
            await UserProgress.updateOne(
                { userId: req.user._id, problemId: revision.problemId },
                {
                    $set: {
                        markedForRevision: false,
                        nextRevisionDate: null
                    }
                }
            );

            await revision.deleteOne();
            res.json({ message: 'Revision removed' });
        } else {
            res.status(404).json({ message: 'Revision not found' });
        }
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

module.exports = {
    getRevisionSchedule,
    getOverdueRevisions,
    markRevisionComplete,
    getRevisionStats,
    deleteRevision
};

File: dsa-tracker-backend/src/main/java/com/dsatracker/controller/RevisionController.java
package com.dsatracker.controller;

import com.dsatracker.exception.BadRequestException;
import com.dsatracker.exception.ResourceNotFoundException;
import com.dsatracker.model.Problem;
import com.dsatracker.model.Revision;
import com.dsatracker.model.UserProgress;
import com.dsatracker.repository.ProblemRepository;
import com.dsatracker.repository.RevisionRepository;
import com.dsatracker.repository.UserProgressRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/revision")
public class RevisionController {

    @Autowired
    private RevisionRepository revisionRepository;

    @Autowired
    private UserProgressRepository userProgressRepository;

    @Autowired
    private ProblemRepository problemRepository;

    private String getAuthenticatedUserId() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal instanceof UserDetails) {
            return ((UserDetails) principal).getUsername();
        }
        throw new BadRequestException("Unauthorized");
    }

    @GetMapping
    public ResponseEntity<?> getRevisionSchedule(@RequestParam(value = "status", required = false) String status) {
        String userId = getAuthenticatedUserId();

        List<Revision> revisions = revisionRepository.findByUserId(userId);

        LocalDate today = LocalDate.now();
        Date todayDate = Date.from(today.atStartOfDay(ZoneId.systemDefault()).toInstant());

        // Filter based on status
        if ("completed".equalsIgnoreCase(status)) {
            revisions.removeIf(r -> !r.isCompleted());
        } else if ("pending".equalsIgnoreCase(status)) {
            revisions.removeIf(Revision::isCompleted);
        } else if ("overdue".equalsIgnoreCase(status)) {
            revisions.removeIf(r -> r.isCompleted() || !r.getScheduledDate().before(todayDate));
        }

        // Sort by scheduled date ascending
        revisions.sort(Comparator.comparing(Revision::getScheduledDate));

        // Populate problemId in output response objects
        List<Map<String, Object>> populated = new ArrayList<>();
        for (Revision r : revisions) {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("_id", r.getId());
            map.put("userId", r.getUserId());
            map.put("scheduledDate", r.getScheduledDate());
            map.put("completed", r.isCompleted());
            map.put("completedAt", r.getCompletedAt());
            map.put("revisionNumber", r.getRevisionNumber());
            map.put("createdAt", r.getCreatedAt());

            Problem p = problemRepository.findById(r.getProblemId()).orElse(null);
            map.put("problemId", p); // Nest full problem info inside problemId key to match populate() on frontend

            populated.add(map);
        }

        return ResponseEntity.ok(populated);
    }

    @GetMapping("/overdue")
    public ResponseEntity<?> getOverdueRevisions() {
        String userId = getAuthenticatedUserId();
        LocalDate today = LocalDate.now();
        Date todayDate = Date.from(today.atStartOfDay(ZoneId.systemDefault()).toInstant());

        List<Revision> revisions = revisionRepository.findByUserIdAndCompletedFalse(userId);
        revisions.removeIf(r -> !r.getScheduledDate().before(todayDate));
        revisions.sort(Comparator.comparing(Revision::getScheduledDate));

        List<Map<String, Object>> populated = new ArrayList<>();
        for (Revision r : revisions) {
            Map<String, Object> map = new LinkedHashMap<>();
            map.put("_id", r.getId());
            map.put("userId", r.getUserId());
            map.put("scheduledDate", r.getScheduledDate());
            map.put("completed", r.isCompleted());
            map.put("completedAt", r.getCompletedAt());
            map.put("revisionNumber", r.getRevisionNumber());
            map.put("createdAt", r.getCreatedAt());

            Problem p = problemRepository.findById(r.getProblemId()).orElse(null);
            map.put("problemId", p);

            populated.add(map);
        }

        return ResponseEntity.ok(populated);
    }

    @PutMapping("/{id}/complete")
    public ResponseEntity<?> markRevisionComplete(@PathVariable("id") String id) {
        String userId = getAuthenticatedUserId();
        Revision revision = revisionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Revision not found"));

        if (!revision.getUserId().equals(userId)) {
            throw new BadRequestException("Unauthorized access to this revision");
        }

        revision.setCompleted(true);
        revision.setCompletedAt(new Date());
        revisionRepository.save(revision);

        // Update UserProgress revision metadata
        UserProgress progress = userProgressRepository.findByUserIdAndProblemId(userId, revision.getProblemId())
                .orElse(null);

        if (progress != null) {
            progress.setRevisionCount(progress.getRevisionCount() + 1);
            if (progress.getRevisionDates() == null) {
                progress.setRevisionDates(new ArrayList<>());
            }
            progress.getRevisionDates().add(new Date());

            // Schedule next revision (number 2 -> 3, etc. with days 1, 7, 30)
            if (revision.getRevisionNumber() < 3) {
                int[] revisionDays = {1, 7, 30};
                int nextDay = revisionDays[revision.getRevisionNumber()]; // revisionNumber is 1-indexed, so 1 -> 7 days, 2 -> 30 days
                Calendar cal = Calendar.getInstance();
                cal.add(Calendar.DAY_OF_YEAR, nextDay);
                progress.setNextRevisionDate(cal.getTime());
            } else {
                progress.setNextRevisionDate(null);
            }

            userProgressRepository.save(progress);
        }

        return ResponseEntity.ok(revision);
    }

    @GetMapping("/stats")
    public ResponseEntity<?> getRevisionStats() {
        String userId = getAuthenticatedUserId();

        LocalDate today = LocalDate.now();
        LocalDate tomorrow = today.plusDays(1);
        Date dToday = Date.from(today.atStartOfDay(ZoneId.systemDefault()).toInstant());
        Date dTomorrow = Date.from(tomorrow.atStartOfDay(ZoneId.systemDefault()).toInstant());

        List<Revision> allUserPendingRevisions = revisionRepository.findByUserIdAndCompletedFalse(userId);

        int dueToday = 0;
        int overdue = 0;
        for (Revision r : allUserPendingRevisions) {
            Date sDate = r.getScheduledDate();
            if (sDate.before(dToday)) {
                overdue++;
            } else if (sDate.after(dToday) && sDate.before(dTomorrow) || sDate.equals(dToday)) {
                dueToday++;
            }
        }

        int totalPending = allUserPendingRevisions.size();

        // Weak Topics Calculation: group in-memory based on progress marked for revision
        List<UserProgress> userProgressList = userProgressRepository.findByUserId(userId);
        userProgressList.removeIf(p -> !p.isMarkedForRevision());

        Map<String, Integer> topicCounts = new HashMap<>();
        for (UserProgress up : userProgressList) {
            Problem p = problemRepository.findById(up.getProblemId()).orElse(null);
            if (p != null && p.getTopics() != null) {
                for (String t : p.getTopics()) {
                    topicCounts.put(t, topicCounts.getOrDefault(t, 0) + 1);
                }
            }
        }

        List<Map<String, Object>> weakTopics = topicCounts.entrySet().stream()
                .sorted((entry1, entry2) -> entry2.getValue().compareTo(entry1.getValue()))
                .limit(3)
                .map(entry -> {
                    Map<String, Object> map = new LinkedHashMap<>();
                    map.put("topic", entry.getKey());
                    map.put("count", entry.getValue());
                    return map;
                })
                .collect(Collectors.toList());

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("dueToday", dueToday);
        stats.put("overdue", overdue);
        stats.put("totalPending", totalPending);
        stats.put("weakTopics", weakTopics);

        return ResponseEntity.ok(stats);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteRevision(@PathVariable("id") String id) {
        String userId = getAuthenticatedUserId();
        Revision revision = revisionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Revision not found"));

        if (!revision.getUserId().equals(userId)) {
            throw new BadRequestException("Unauthorized access to this revision");
        }

        UserProgress progress = userProgressRepository.findByUserIdAndProblemId(userId, revision.getProblemId())
                .orElse(null);

        if (progress != null) {
            progress.setMarkedForRevision(false);
            progress.setNextRevisionDate(null);
            userProgressRepository.save(progress);
        }

        revisionRepository.delete(revision);

        Map<String, String> response = new HashMap<>();
        response.put("message", "Revision removed");
        return ResponseEntity.ok(response);
    }
}
